### ISSUE TITLE
Handle exceptions in StepBody

### ISSUE THREAD

**Issue by ccpony** [danielgerlag/workflow-core#9]
Daniel - I'm not clear about exception handling in stepbodies.  An exception thrown in a stepbody sets the workflow to status 0 and attempts re-tries continuously.  The exception is logged to the console but not reflected back to the UI.  Nor is there any record of the error kept in the database.  What is best practice for handling an exception within the stepbody so that the user is alerted and the information can be conveyed back to technical support?

Corollary to this question: how does one response to an stepbody exception by terminating the workflow (and, hopefully, alerting the developers).

Thanks.

P.S.  I'm spending a lot of time with this application now.  Please direct me as to how I can help.

**Comment by danielgerlag** [comment]
I'm thinking of adding an .OnException() method to the step builder.  This way you would be able to specify the exception handling per step, ie. either retry sometime in the future, suspend the workflow or terminate the workflow.  I could also add an OnException event handler to the WorkflowHost service.
Would these proposals solve your issue?

If you can hang in there a few days, I can make these changes and publish a new version, or if you are up to getting more involved in the project, I'm happy for you to create a pull request.

It should also be logging uncaught exceptions to the db, I will look into this, and write some tests around it - you could also help by looking into this if you want.

**Comment by ccpony** [comment]
Hello Daniel.

Yes.  Adding OnException() into the stepbuilder would be very helpful - - and definitely adding the OnException event to the Workflow service itself is essential to reduce the amount of error handling code, if possible.

I looked into doing this myself, but I’m still getting familiarize with your code.  It impresses me as a significant enhancement so I’ll leave this one up to you.

Let me ask you this: where should the uncaught exceptions be logged in the database?  I looked but could not find that.  Let me know and I should be able to debug that area, if need be.

Thanks.

Jeff
